File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        def canEatAllBananas(speed):
            kokoLoc = 0
            timeRemaining = h

            while kokoLoc < len(piles) and timeRemaining > 0:
                pile = piles[kokoLoc]
                timeRemaining -= ceil(pile / speed)

                if timeRemaining >= 0:
                    kokoLoc += 1
                else:
                    break

            return kokoLoc == len(piles)

        l, r = 1, max(piles)
        minSpeed = r

        while l <= r:
            mid = l + ((r-l)//2)

            if canEatAllBananas(mid):
                r = mid - 1
                minSpeed = min(mid, minSpeed)

            else:
                l = mid + 1

        return minSpeed

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.go
func minEatingSpeed(piles []int, h int) int {

    maxPile := 0
    for _, val := range piles {
        if val > maxPile { maxPile = val }
    }

    lowerBound := (len(piles) + h - 1) / h
    upperBound := maxPile * len(piles)

    checkSpeed := func(speed int) bool {
        hoursRemaining := h
        for i := 0; i < len(piles); i++ {
            if hoursRemaining <= 0 {
                return false
            }
            hoursRemaining -= (piles[i] + speed - 1) / speed
        }
        if hoursRemaining < 0 {
            return false
        } else {
            return true
        }
    }

    for lowerBound <= upperBound {
        mid := lowerBound + ((upperBound - lowerBound) / 2)
        if checkSpeed(mid) {
            upperBound = mid - 1
        } else {
            lowerBound = mid + 1
        }
    }

    return lowerBound
}